examples/basic/gpu_shared_memory.c

```diff
@@ -0,0 +1,245 @@
+/*
+ *
+ * Copyright (C) 2024 Intel Corporation
+ *
+ * Under the Apache License v2.0 with LLVM Exceptions. See LICENSE.TXT.
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+ *
+ */
+
+#include <assert.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <ze_api.h>
+
+#include "umf/memory_pool.h"
+#include "umf/pools/pool_disjoint.h"
+#include "umf/providers/provider_level_zero.h"
+
+// This function initializes the Level Zero driver and creates a context in the
+// first GPU device found.
+int level_zero_init(ze_driver_handle_t *driver, ze_device_handle_t *device,
+                    ze_context_handle_t *context) {
+
+    // Initialize the Level Zero driver
+    ze_result_t ze_result = zeInit(0);
+    if (ze_result != ZE_RESULT_SUCCESS) {
+        fprintf(stderr, "Failed to initialize Level Zero!\n");
+        return -1;
+    }
+
+    // Discover all the driver instances
+    uint32_t driverCount = 0;
+    ze_result = zeDriverGet(&driverCount, NULL);
+    if (ze_result != ZE_RESULT_SUCCESS) {
+        fprintf(stderr, "Failed to get number of Level Zero drivers!\n");
+        return -1;
+    }
+    if (driverCount == 0) {
+        fprintf(stderr, "There are no Level Zero drivers on the system!\n");
+        return -1;
+    }
+
+    ze_driver_handle_t *all_drivers =
+        (ze_driver_handle_t *)calloc(driverCount, sizeof(ze_driver_handle_t));
+    if (all_drivers == NULL) {
+        fprintf(stderr, "Failed to allocate memory for the list of Level Zero "
+                        "drivers!\n");
+        return -1;
+    }
+
+    ze_result = zeDriverGet(&driverCount, all_drivers);
+    if (ze_result != ZE_RESULT_SUCCESS) {
+        fprintf(stderr, "Failed to get list of Level Zero drivers!\n");
+        goto drivers_destroy;
+    }
+
+    // Find a driver instance with a GPU device
+    ze_device_handle_t *all_devices = NULL;
+    for (uint32_t i = 0; i < driverCount; ++i) {
+        if (all_drivers[i] == NULL) {
+            fprintf(stderr, "Error - driver handle can't be NULL!\n");
+            goto drivers_destroy;
+        }
+
+        // Discover all the devices
+        uint32_t deviceCount = 0;
+        ze_result = zeDeviceGet(all_drivers[i], &deviceCount, NULL);
+        if (ze_result != ZE_RESULT_SUCCESS) {
+            fprintf(stderr, "Failed to get number of Level Zero devices!\n");
+            goto drivers_destroy;
+        }
+        if (deviceCount == 0) {
+            fprintf(stderr, "There are no Level Zero devices on the system!\n");
+            goto drivers_destroy;
+        }
+
+        all_devices = (ze_device_handle_t *)calloc(deviceCount,
```

Comment:
Ensure that you free this variable, in case of error

Reply:
IMO it is released in case of errors - do you see any real problem here?

Reply:
I switched to using utils file provided in #418